### PR TITLE
Fix duration buckets overflow

### PR DESCRIPTION
**Pull request by roman-mazhut**
Fix for #264 
This PR addresses an overflow issue in `ExponentialDurationBuckets`, which can cause negative durations due to integer wrap-around when converting large values into `time.Duration`.

### Problem
Under the hood, `time.Duration` is represented as an `int64` nanosecond value. When the exponential scaling factor produces a number too large for `int64`, an overflow occurs, resulting in buckets with negative durations on some architectures.

Example:
```
tally.ExponentialDurationBuckets(time.Second, 1.3, 90)
```
Results in:
**x86**: `[..., -2562047h47m16.854775808s, -2562047h47m16.854775808s]`
**ARM**: `[..., 2562047h47m16.854775807s, 2562047h47m16.854775807s]`
(ARM seems to perform saturating calculations instead of wrapping.)

### Solution
This PR introduces a `safeFloat64ToInt64` function, which checks for overflows before converting to `time.Duration`. If an overflow is detected, it returns `math.MinInt64` or `math.MaxInt64`, preventing undefined behavior.

Now, `ExponentialDurationBuckets` will still return identical buckets in case of overflow, but the behavior will be consistent across different CPU architectures.

## Summary by CodeRabbit

- **New Features**
  - Enhanced the logic for generating duration buckets, improving precision and handling of extreme input values.

- **Bug Fixes**
  - Added tests to ensure that the system behaves correctly under maximum limit conditions, preventing unexpected errors.

- **Tests**
  - Introduced new test functions to validate the behavior of duration calculations and conversions, enhancing overall test coverage.

- **Chores**
  - Updated the workflow configuration to support testing across multiple operating systems, improving testing flexibility.

### CODE REVIEW

**Review by prateek — COMMENTED** [review]
hey thanks for starting this off, few different things to change up 

- [ ] do the bounds check on all the bucket ctors (linear/exponential/etc) for both bucket types (duration/values)
- [x] turn on CI for both x86+arm -- https://docs.travis-ci.com/user/multi-cpu-architectures/ seems like it should be quite tractable

File: histogram_test.go [hunk 1]
```diff
@@ -166,6 +166,14 @@ func TestMustMakeExponentialDurationBucketsPanicsOnBadFactor(t *testing.T) {
 	})
 }
 
+func TestMustMakeExponentialDurationBucketsOverflow(t *testing.T) {
```
Comment (prateek) [inline histogram_test.go:169]: add a test to capture the expected generated buckets on overflow

File: histogram.go [hunk 2]
```diff
@@ -368,7 +368,7 @@ func ExponentialDurationBuckets(start time.Duration, factor float64, n int) (Dur
 	curr := start
 	for i := range buckets {
 		buckets[i] = curr
-		curr = time.Duration(float64(curr) * factor)
+		curr = time.Duration(safeFloat64ToInt64(float64(curr) * factor))
```
Comment (prateek) [inline histogram.go:371]: this can end up generating invalid buckets. e.g. if you're generating 10 buckets and bucket 7 is already overflowing, the last 3 buckets are going to be redundant/duplicates; this will cause issues on the metric consumption side (`histogramPercentile` in m3ql will complain about duplicate bucket ranges) 

i.e. we should probably pick a semantic in terms of either truncating to generate only valid, non-duplicate buckets OR we should consider raising a panic when there's invalid bucket definitions. not convinced which is better - but we should pick a path in this diff.

Comment (roman-mazhut) [reply histogram.go:371]: I thought about removing duplicate buckets, but my concern is that the third argument of the `ExponentialDurationBuckets` function is the number of buckets so that some clients may rely on that number.
Also, raising a panic is not a backward-compatible change.

File: histogram.go [hunk 3]
```diff
@@ -368,7 +369,7 @@ func ExponentialDurationBuckets(start time.Duration, factor float64, n int) (Dur
...
```
Comment (prateek) [inline histogram.go:372]: options we talked about offline, consider the existing signature 
```
func ExponentialDurationBuckets(start time.Duration, factor float64, n int) (DurationBuckets, error) {
```

we could change this to
```
func ExponentialDurationBuckets(
  start time.Duration, 
  factor float64, 
  n int,
  opts ...BucketOpts,
) (DurationBuckets, error) {
...
}

// you can have opts for "autofix buckets" or "error on invalid bucket range/overflow/etc
```

**Review by prateek — APPROVED** [review]
